### ISSUE TITLE
Train detection with digilines - Implement "mesecon" and "digiline" modes for ATC rails

### ISSUE THREAD

**Issue by gbl08ma** [orwell96/advtrains#3]
I have added "detector rails" to other Minetest train mods, and I'd like to have train detection support in this mod too so I could switch my worlds to it, as it is vastly superior than other cart mods.

However, I gave the source code a look and I don't understand how the train-track connection works. It seems that trains predict their paths using `pathpredict`, and that whenever the possible paths change (e.g. when a switch is right-clicked), these predictions are invalidated with `invalidate_all_paths`. Ideally I'd like the track to "know" when a train passes over it, so it could send a digiline message. However, it doesn't seem like the track components can have any idea that a train is going over them.

The other way to do this would be to register a globalstep function which would work similarly to the damage function in damage.lua, which would regularly check whether a train is near a detector track using `minetest.get_objects_inside_radius`, and have that piece of track send a digiline message if a train is inside the radius. But previous experience with using this approach on other mods told me that this is unreliable: a short train going at high speed may go through a detector track undetected, because on one step, the train is still too far "before" the detector, and the step after that, it has completely gone through the detector and is now too far ("after" the detector) to trigger a detection. The problem gets worse if there are many trains or the server is slow for some reason. This can be mitigated by increasing the radius of detection, but to me this feels too much like a hack and furthermore it may work incorrectly if two trains happen to be in vicinity of the detector.

As to why digilines and not just mesecons, the reason is simple: with digilines the detector could send information about the train passing through, including:
- The number of wagons
- The speed and direction
- The type of train (electric, steam...) and, if that is ever implemented, its fuel level (for steam and diesel locomotives)
- The train ID, if it ever becomes possible to assign names/IDs to trains or just read the existing internal IDs (i.e. the indices of the `advtrains.trains` table)

With mesecons the best it could do would be to turn on an emitter, losing all additional information. Of course, there's no reason not to support both, I'm just stating why digilines would work better for me because it can convey additional information.
In my experience it's also a bit easier to send a digiline message than to develop a Mesecon emitter, because the latter, as far as I know, requires using different nodes for the "on" and "off" states of the detector.

**Comment by orwell96** [comment]
The next thing i will do is making switches controllable by mesecons. If you like, I'll make detector rails afterwards. I know my source code best, and you are right when you say that  abms and globalsteps alone are the wrong approach. I already have an idea on how to implement it.
When I get to implementing Automatic Train Control rails, there will be a digiline interface how you imagine it.

**Comment by gbl08ma** [comment]
That sounds good. It would make sense to have simpler components interfacing with mesecons, and the "smarter" ATC components working with digilines.

A problem with both is that even though trains can travel through unloaded map chunks, mesecons and digilines blocks like luacontrollers will not work if the chunk is unloaded, so the ATC system would stop working correctly as the players walk away. Perhaps this could be fixed by force-loading the chunks where trains are going, but it could increase the server load too much.

**Comment by orwell96** [comment]
known problem. I will make them just doing nothing when player's away. If someone needs, he can use forceload anchors.
BTW trains do not collide when they are unloaded.
BTW2 unloaded mapchunk travelling does not work for some reason ATM.

**Comment by orwell96** [comment]
Moving forward...
We have detector rails and the needed underlying system now, only a small step to ATC.

**Comment by gpcf** [comment]
I don't think this is the right way to implement ATC. After all, you want your trains to run through unloaded mapchunks, which is a messy process. I would implemented using an external program that controls the trains. That program would contain a simutrans-style algorithm that sends the trains to their destination and sets the switches and signals accordingly. Would be a lot simpler than implementing this in mesecons, which would lead to a lot of issues with loaded/unloaded chunks. I would write that program in C for performance reasons.

I currently do not have the time to implement this myself, but I may do this if I have time in the future.

**Comment by orwell96** [comment]
The primary intended use is having such a rail before a signal, which then tells the train to stop when the signal is red. Of course it will offer far more possibilities, and together with forceload anchors in stations(there are mods for this) one could build such automated stuff
However, @gpfc's idea is good if you want to properly automate everything. @gpcf, feel free to fork and do it. I would really like to see this, but I will stay focused on the 'core'.

**Comment by gbl08ma** [comment]
On the one hand, @gpcf is right, on the other hand, wrong. Let me explain: it's true that digilines+mesecons alone will never work right to manage bigger parts of the train network as a whole, because they don't work on unloaded mapchunks. This makes even "simple" things like making a system to ensure even spacing between trains impossible. An external program (even just another Lua mod would suffice) would need to take care of this.

Ideally, advtrains would provide some sort of API that lets these external systems query the current position of trains (independently of their mapchunks being loaded or not) and actuate on the world (by changing switches, signals, displays in stations, crossing barriers and what else may be added in the future). If you want to make it more realistic, you can make it so that the external programs can only know through what detectors trains passed. (If you want to plan for the far far future, you can even think of stuff like "GPS receiver" add-ons for trains that would let those external systems query their position at all times, etc. but that's irrelevant now).
Note that this "API approach" has the benefit that all the work of the ATC is offloaded to another mod, especially good if @orwell96 wants to stay focused on the core.

Now, I think there's value in having digilines and mesecons compatibility for the simple stuff @orwell96 talked about. Want to make it so that a switch changes direction every time a train passes through? A luacontroller is enough for that (as long as you don't mind the switch not changing when nobody's "watching"). What about playing some noteblock sound when a train passes? Yup, digilines+mesecons is good enough. And the list of simple stuff goes on.

**TL;DR: offload complex ATC to another mod, allow simple stuff to be made using in-world tools (i.e. this mod, digilines and/or mesecons).**

I have not updated my copy of advtrains in a long time, I'll do so when I can, to test the new detector rails.

**Comment by gbl08ma** [comment]
The current Lua ATC capabilities are already very interesting, but in my map they aren't of much use (at least, not yet). Going by the docs, I have noticed that there's no way for the Lua ATC components to interact with non-advtrains components like LuaControllers, other than the ability to toggle Mesecon switches.

Consider the following use case: I want to display a ETA for the next train on each subway station. Right now, I have a prototype implemented in one station which works based on the advtrains detectors: when a train enters or leaves a station, LuaControllers receive the Mesecon signal from the detectors, send some messages through digilines to the display controller for that station, which then does some calculations to come up with the ETA for the next train on that line, and sends it to the station displays.
This works relatively well, but only as long as there's someone on the station to keep its chunks loaded so that all the mesecon and digilines stuff works.

Since Lua ATC rails, unlike LuaControllers and the like, are meant to always work even in unloaded chunks, I would like to move most of the ETA calculation math to them. That way, they would never miss a train passing over them, and we could have an always up-to-date ETA for each station stored in the S table of the Lua ATC environment. Until here, I think all the necessary parts are already implemented - the problem is what comes next:

Since the ETAs would now be stored in the Lua ATC environment(s), LuaControllers need some way to access those estimates, once they "wake up", i.e. when their chunk is loaded. The solution that would be most convenient to me would be to make the Lua ATC rails "talk" digilines, so that LuaControllers can send them messages and ATC rails can reply. This would allow for sharing information between the ATC environments and any LuaController-based systems.
However, I can imagine other systems: for example, a special block (not rails) that's specifically meant to be connected to digilines and can access a Lua ATC environment, or perhaps just its S table.

What do you think?

**Comment by gpcf** [comment]
You probably want to have a display block that shows the time that the train will take. This should be relatively easy to implement just with the Lua ATC rails

**Comment by gbl08ma** [comment]
Did you read what I wrote carefully? I want to integrate Advtrains with the display system I already have, which uses LuaControllers and Digilines. In this example, the time at the bottom is the ETA in MM:SS:

![display example](https://cloud.githubusercontent.com/assets/984584/22629786/b863927a-ebe4-11e6-91c2-0f8ff6ec4511.png)

(that display was made by me and is not released as a standalone mod yet, but it works more or less the same way as the LCD that comes with Digilines)

I wouldn't disagree with implementing a display block that's specific to Advtrains and just shows the ETA, it would make it easier for people who don't want to implement a full "smart" display system like mine. But that wouldn't solve my problem, since I'd need to use two displays then, one from Advtrains and the other for everything else... furthermore, these displays are meant to show other stuff. For example, when a train is approaching the station, they show a message like "Train with X cars" so that in the case of shorter trains players have time to move to the far side of the platform before the train leaves.

**Comment by codexp** [comment]
any news?

TL;DR Is there any option to detect passing trains?

**Comment by gpcf** [comment]
Eugen Wesseloh <notifications@github.com> writes:

> any news?
>
> TLDR; Is there any option to detect passing trains?
LuaATC. Its main advantage is working in unloaded blocks. It works quite
well, probably better than any digiline or mesecon based system could work.

**Comment by codexp** [comment]
but it is limited to ATC commands?

*EDIT*
oh, sorry LuaATC, where can I find API doc for LuaATC?

*EDIT2*
Lua Automation doc seems to be in https://github.com/orwell96/advtrains/tree/master/advtrains_luaautomation

**Comment by orwell96** [comment]
Won't implement. Existing possibilities are sufficient, and the conflict luaded vs. unloaded will always remain
